libs/blockchain/src/capacity-checker.service.ts

```diff
@@ -11,37 +11,65 @@
  *          - throw an error if an empty value is encountered
  */
 import { ICapacityLimit } from '#types/interfaces/capacity-limit.interface';
-import { Injectable, Logger } from '@nestjs/common';
+import { Inject, Injectable, Logger, OnApplicationBootstrap, OnModuleDestroy } from '@nestjs/common';
 import { InjectRedis } from '@songkeys/nestjs-redis';
 import { Redis } from 'ioredis';
 import { EventEmitter2 } from '@nestjs/event-emitter';
 import { ICapacityInfo } from './types';
-import { ConfigService } from '@nestjs/config';
-import { IBlockchainConfig } from './blockchain.config';
+import blockchainConfig, { IBlockchainConfig } from './blockchain.config';
 import { BlockchainRpcQueryService } from './blockchain-rpc-query.service';
 
 export const CAPACITY_EXHAUSTED_EVENT = 'capacity.exhausted';
 export const CAPACITY_AVAILABLE_EVENT = 'capacity.available';
 
 const EPOCH_CAPACITY_PREFIX = 'epochCapacity:';
+const CURRENT_CHAIN_CAPACITY_KEY = 'currentChainCapacity:';
+const CAPACITY_CHECK_INTERVAL_MS = 10000; // 10 seconds
 
 @Injectable()
-export class CapacityCheckerService {
+export class CapacityCheckerService implements OnApplicationBootstrap, OnModuleDestroy {
   private readonly logger: Logger;
 
   private lastCapacityEpoch: number;
 
   private lastCapacityUsedCheck: bigint;
 
+  private capacityCheckInterval: NodeJS.Timeout;
+
+  async onApplicationBootstrap() {
+    await this.updateCachedCapacity();
+  }
+
+  onModuleDestroy() {
+    if (this.capacityCheckInterval) {
+      clearInterval(this.capacityCheckInterval);
+    }
+  }
+
   // eslint-disable-next-line no-useless-constructor
   constructor(
     private readonly blockchainService: BlockchainRpcQueryService,
-    private readonly configService: ConfigService,
+    @Inject(blockchainConfig.KEY) private readonly config: IBlockchainConfig,
     @InjectRedis() private readonly redis: Redis,
     private readonly eventEmitter: EventEmitter2,
-    // eslint-disable-next-line no-empty-function
   ) {
     this.logger = new Logger();
+
+    // NOTE: Why we do this on an interval instead of using an expiring cache key & fetching inside checkForSufficientCapacity:
```

Comment:
I might be wrong but I don't think setting the cache on Redis and querying it for every request would be that much of a overhead. Calling chain RPC is more expensive but I assume Redis calls should be really fast. Do we have any measurement to support that this is necessary?

Reply:
The idea here is that, in the case where the key has expired & the capacity needs to be fetched from the chain, we don't want that RPC call in the critical path of submitting a transaction. Better to have it update in a separate async task. (Yes, I realize that given the single-threaded nature of Node this may be a fool's errand, but at least I feel better about it 😅 )

Reply:
There's still one capacity check in the critical path. The cached check just checks to see if we've exceeded our configured limits.

There's a per-transaction check that makes sure we actually have enough capacity for this specific extrinsic, that actually calls the chain. We want to get this out of the critical path as well, and then catch any out-of-capacity errors and re-queue, but this PR is big enough already. Tracking in #760